core/commands/files/files.go

```diff
@@ -0,0 +1,555 @@
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	gopath "path"
+	"strings"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	core "github.com/ipfs/go-ipfs/core"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	mfs "github.com/ipfs/go-ipfs/mfs"
+	path "github.com/ipfs/go-ipfs/path"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+var log = u.Logger("cmds/files")
+
+var FilesCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Manipulate unixfs files",
+		ShortDescription: `
+Files is an API for manipulating ipfs objects as if they were a unix filesystem.
+`,
+	},
+	Subcommands: map[string]*cmds.Command{
+		"read":  FilesReadCmd,
+		"write": FilesWriteCmd,
+		"mv":    FilesMvCmd,
+		"cp":    FilesCpCmd,
+		"ls":    FilesLsCmd,
+		"mkdir": FilesMkdirCmd,
+		"stat":  FilesStatCmd,
+		"rm":    FilesRmCmd,
+	},
+}
+
+var FilesStatCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "display file status",
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to node to stat"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[0]
+		fsn, err := mfs.Lookup(node.FilesRoot, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		nd, err := fsn.GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		k, err := nd.Key()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		res.SetOutput(&Object{
+			Hash: k.B58String(),
+		})
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*Object)
+			return strings.NewReader(out.Hash), nil
+		},
+	},
+	Type: Object{},
+}
+
+var FilesCpCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "copy files into mfs",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("src", true, false, "source object to copy"),
+		cmds.StringArg("dest", true, false, "destination to copy object to"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		src := req.Arguments()[0]
+		dst := req.Arguments()[1]
+
+		var nd *dag.Node
+		switch {
+		case strings.HasPrefix(src, "/ipfs/"):
+			p, err := path.ParsePath(src)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			obj, err := core.Resolve(req.Context(), node, p)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd = obj
+		default:
+			fsn, err := mfs.Lookup(node.FilesRoot, src)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			obj, err := fsn.GetNode()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd = obj
+		}
+
+		err = mfs.PutNode(node.FilesRoot, dst, nd)
```

Comment:
valid as in, 'contains valid unix path elements' ? or valid as in 'this path exists'.

The latter is handled inside `PutNode`